File: solutions/24. Swap Nodes in Pairs.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* swapPairs(struct ListNode* head) {
  // make the initial case not special
  struct ListNode sent = {.val = 0, .next = head};

  struct ListNode* prev = &sent;
  struct ListNode* curr = prev->next;

  /*
   * In each loop,
   * Initial:
   *    0 1 2 3 ...
   *  prev: 0
   *  curr: 1
   * End:
   *    0 2 1 3 ...
   *  prev: 1
   *  curr: 3
   */
  while (curr && curr->next) {
    prev->next = curr->next;        // 0 -> 2
    curr->next = curr->next->next;  // 1 -> 3
    prev->next->next = curr;        // 2 -> 1

    prev = curr;        // 0 = 1
    curr = curr->next;  // 1 == 3
  }

  return sent.next;
}

File: solutions/24. Swap Nodes in Pairs.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn swap_pairs(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        // An elegant solution from
        // https://leetcode.com/problems/swap-nodes-in-pairs/solutions/2882193/simple-rust-recursive-solution/?orderBy=most_votes&languageTags=rust.
        match head {
            Some(mut curr) => match curr.next {
                Some(mut next) => {
                    curr.next = Self::swap_pairs(next.next);
                    next.next = Some(curr);
                    Some(next)
                }
                None => Some(curr),
            },
            None => head,
        }
    }
}
